src/modules/dashboard/components/Dialogs/AddExistingSafeDialog/AddExistingSafeDialog.tsx

```diff
@@ -1,45 +1,268 @@
+import { defineMessages, useIntl } from 'react-intl';
+import { isAddress } from 'web3-utils';
 
 import Dialog, { DialogProps, ActionDialogProps } from '~core/Dialog';
 import { ActionForm, SelectOption } from '~core/Fields';
-import { SAFE_NETWORKS } from '~modules/constants';
+import {
+  FETCH_ABORTED,
+  SAFE_NAMES_MAP,
+  SAFE_NETWORKS,
+} from '~modules/constants';
 import { Address } from '~types/index';
 import { ActionTypes } from '~redux/index';
 import { pipe, withMeta, mapPayload } from '~utils/actions';
 import { WizardDialogType } from '~utils/hooks';
 
 import DialogForm from './AddExistingSafeDialogForm';
+import { getTxServiceBaseUrl } from '~modules/dashboard/sagas/utils/safeHelpers';
 
 export interface FormValues {
   chainId: string;
   contractAddress: Address;
-  safeName: string;
-  annotation: string;
+  moduleContractAddress?: Address;
+  safeName?: string;
+  annotation?: string;
 }
 
 type Props = Required<DialogProps> &
   WizardDialogType<object> &
   ActionDialogProps;
 
+export const MSG = defineMessages({
+  contractAddressError: {
+    id: 'dashboard.AddExistingSafeDialog.contractAddressError',
+    defaultMessage: 'Please enter a contract address',
+  },
+  safeNameError: {
+    id: 'dashboard.AddExistingSafeDialog.safeNameError',
+    defaultMessage: 'Please enter a safe name',
+  },
+  moduleAddressError: {
+    id: 'dashboard.AddExistingSafeDialog.moduleAddressError',
+    defaultMessage: 'Please enter a module address',
+  },
+  safeAlreadyExistsError: {
+    id: 'dashboard.AddExistingSafeDialog.CheckSafe.alreadyExists',
+    defaultMessage: 'Safe already exists in this colony.',
+  },
+  fetchFailedError: {
+    id: 'dashboard.AddExistingSafeDialog.CheckSafe.fetchFailed',
+    defaultMessage: `Could not fetch {type} details. Please check your connection and try again.`,
+  },
+  safeNotFoundError: {
+    id: 'dashboard.AddExistingSafeDialog.CheckSafe.safeCheck',
+    defaultMessage: `Safe not found on {selectedChain}`,
+  },
+  moduleNotConnectedError: {
+    id: 'dashboard.AddExistingSafeDialog.CheckSafe.safeCheck',
+    defaultMessage: `Module not connected to safe on {selectedChain}`,
+  },
+  moduleNotFoundError: {
+    id: 'dashboard.AddExistingSafeDialog.CheckSafe.safeCheck',
+    defaultMessage: `Module not found on {selectedChain}`,
+  },
+});
 const displayName = 'dashboard.AddExistingSafeDialog';
 
+// "this" is needed within yup.test().
+/* eslint-disable react/no-this-in-sfc */
 const AddExistingSafeDialog = ({
   colony: { colonyAddress, colonyName, safes },
   callStep,
   prevStep,
   cancel,
   close,
 }: Props) => {
+  const { formatMessage } = useIntl();
   const history = useHistory();
+  const loadingSafeState = useState<boolean>(false);
+  const loadingModuleState = useState<boolean>(false);
+  const [abortController, setAbortController] = useState<
+    AbortController | undefined
+  >(undefined);
+  const stepState = useState<number>(1);
```

Comment:
Not saying this is wrong or anything but why do you not destructure the array returned by `useState`? You refer to the state value later by using `stepState[0]` which could be less readable than something like `stepIndex`. And you end up destructuring this array in the `AddExistingSafeDialogForm` component anyway

Reply:
Another good idea